add explicit wait to AsyncCollectiveTensor

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97945
* #97924

This PR adds an explicit wait API to handle the case where if the traced
function only have a single collective with no consumer, this removes
the getattr node and correctly insert wait node

wanchaol committed:
add explicit wait to functional collective

This PR adds an explicit wait API to handle the case where if the traced
function only have a single collective with no consumer, this removes
the getattr node and correctly insert wait node

[ghstack-poisoned]

wanchaol committed:
Update on "add explicit wait to functional collective"


This PR adds an explicit wait API to handle the case where if the traced
function only have a single collective with no consumer, this removes
the getattr node and correctly insert wait node

[ghstack-poisoned]

mrshenli commented on torch/distributed/_spmd/distribute.py:
```diff
@@ -88,7 +88,7 @@ def _dispatch_with_local_tensors(
     def redistribute(arg: object) -> object:
         return (
-            _redistribute_with_local_tensor(arg, *specs[arg])  # type: ignore[index]
+            _redistribute_with_local_tensor(arg, *specs[arg]).wait()  # type: ignore[index]
```
will this wait operator get traced into graph?

I assume this is already the latest time wait can be called because we gonna need the args for this compute right away?

wanchaol replied:
Yes, after adding the `wait()` call it should be traced into the graph, since this is already where we need to allreduce, it's basically the latest point we need the args, which I think it should be the correct behavior. 

Since `_redistribute_with_local_tensor` always returns AsyncCollectiveTensor. I'm not sure why the previous graph works, probably because we still execute the `tensor_constant` get inlined.

wanchaol committed:
Update on "add explicit wait to functional collective"


This PR adds an explicit wait API to handle the case where if the traced
function only have a single collective with no consumer, this removes
the getattr node and correctly insert wait node

[ghstack-poisoned]

wanchaol commented:
@pytorchbot rebase

wanchaol committed:
Update on "add explicit wait to AsyncCollectiveTensor"


This PR adds an explicit wait API to handle the case where if the traced
function only have a single collective with no consumer, this removes
the getattr node and correctly insert wait node

[ghstack-poisoned]
